### DISCUSSION TITLE
Splitting a commit

### DISCUSSION THREAD

**Discussion by ribeiropaulor** [jesseduffield/lazygit#2840] [Q&A] [answered]
I would like to know how to achieve the same result as described in the git book's section 
[Splitting a Commit](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_splitting_a_commit).

So far, the only thing I don't how to do in lazygit is reset the HEAD^.

```bash
git reset HEAD^
```

So far, I am doing the the following steps.

1. [lazygit] In lazygit, I edit the commit I want to split (press e on it).
2. [terminal] Then I exit from lazygit, and in the terminal I run `git reset HEAD^` .
3. [lazygit] Back to lazygit, I stage the changes for "commit-1" and commit it. Then I stage the changes for "commit-2" and commit it. I could split in more than one.
4. [lazygit] Once I have committed all the changes, I continue the rebase (press m, than choose continue... or press c directly).

Done! I end up having the original commit split in "commit-1" and "commit-2".

The only thing that I don't how to do in lazygit, in this case, is `git reset HEAD^` right after I edit the commit.

**Comment by mark2185** [comment]
Please take a look at the [introductory video](https://m.youtube.com/watch?v=CPLdltN7wgE), I think patch building is demonstrated there, around the 8 minute timestamp.

**Reply by ribeiropaulor** [reply]
Thank you!

**Comment by stefanhaller** [comment, marked as answer]
> So far, the only thing I don't how to do in lazygit is reset the HEAD^.
> 
> ```shell
> git reset HEAD^
> ```

This is very easy in lazygit: navigate one line down, hit `g` to bring up the "reset" menu, and hit enter (or `s`) to do a soft reset.

But as Luka said, there are better ways to do such an operation in lazygit: go into the commit that you want to split (by hitting enter on it), "stage" some files or hunks into a custom patch, bring up the custom patch menu (ctrl-p) and choose either "move to new commit" or "move to index".

**Reply by ribeiropaulor** [reply]
Thank you!

**Reply by bradphelan** [reply]
I don't see either "move to new commit" or "move to index"

![image](https://github.com/jesseduffield/lazygit/assets/17650/39c56330-2e9e-4ec6-bd69-e6f4e4e4462a)

**Reply by bradphelan** [reply]
My bad. I was building the patch in the wrong panel. You need to build the patch from the lower left *reflog* panel not by selecting the branch in the middle panel.

**Reply by stefanhaller** [reply]
Then you are in the middle of a rebase. These commands need to perform a rebase for you under the hood, so they are only available when you are not already rebasing.

(Note to self: we should change this so they are always visible, but struck through if not available, with a tooltip explaining why.)

**Reply by bradphelan** [reply]
No I was not in the middle of a rebase

![image](https://github.com/jesseduffield/lazygit/assets/17650/b11d4e2c-8dc8-4c5a-bdbc-86b534fc0b2e)

and

![image](https://github.com/jesseduffield/lazygit/assets/17650/58047050-76b3-4197-a7e5-a56bf3bb1762)

Essentially the same view but different. One view is navigated to from the middle left pane and the other from the bottom left pane. Should there be a difference or not?

**Reply by stefanhaller** [reply]
Ah ok. Well, if you create a patch from the middle panel, then the commit that you are viewing could be on any branch, not necessarily the current one. If it's a commit on some other branch, we'd have to check out that branch first in order to do an interactive rebase on it; we don't support this, and only support these commands for commits of the current branch for simplicity.

Now, we could be smarter and check if the commit that you are viewing is part of the current branch, and enable the commands in that case. However, I think it makes sense to enable those commands only when you are working in the lower left panel, since that's your current branch, and it makes everything clearer (I think).

(Wow, just discovered an edge case while thinking about this: when showing all branches (by typing `<c-l>` and choosing "Toggle show whole git graph"), it's possible to build a patch from a commit on a different branch, and those commands are enabled in that case. They fail with a hard to understand error, so we should fix this.)

**Reply by stefanhaller** [reply]
See #3608.

**Reply by T3sT3ro** [reply]
It would be nice if splitting commit had some better UX/UI. What I would ideally want would be:
- a way to split a commit message as well
- ability to order them (if split patch comes before or after the patch it was split from), like split up/split down
